addons/web/models/models.py

```diff
@@ -153,6 +153,30 @@ def web_read(self, specification: Dict[str, Dict]) -> List[Dict]:
                     for values in values_list:
                         values[field_name] = [x2many_data.get(id_) or {'id': id_} for id_ in values[field_name]]
 
+            elif 'fields' in field_spec and field.type in ('reference', 'many2one_reference'):
+                values_by_id = {
+                    vals['id']: vals
+                    for vals in values_list
+                }
+                for record in self:
+                    if not record[field_name]:
+                        continue
+
+                    if field.type == 'reference':
+                        co_record = record[field_name]
+                    else:  # field.type == 'many2one_reference'
+                        co_record = self.env[record[field.model_field]].browse(record[field_name])
+
+                    if 'context' in field_spec:
+                        co_record = co_record.with_context(**field_spec['context'])
+
+                    values_by_id[record.id][field_name] = co_record.web_read(field_spec['fields'])[0]
```

Comment:
This is not correct, as `web_read()` may legitimately return an empty list.
You have three cases, in fact:
* `record[field_name]` is falsy
* `record[field_name]` is non-falsy but the corresponding record does not exist: it should be handled as if the value was falsy
* `record[field_name]` is non-falsy and the corresponding record exists

The second case is unavoidable, since the value is not a foreign key, and the framework does not check that for you for performance reasons.  In the best case (the value exists), it costs an extra SQL query for nothing.